File: electron/gemini/worksheet-ocr.ts
import { GoogleGenAI } from "@google/genai";
import {
  HandwritingVerificationSchema,
  ReviewSchema,
  SESSION_SCHEMA_VERSION,
  TranscriptionSchema,
  toGeminiJsonSchema,
  type HandwritingVerification,
  type Review,
  type Transcription,
} from "../../shared/contracts";
import { gradeReview } from "../grading/grade-problem";
import { TRANSCRIPTION_PROMPT, verificationPrompt } from "./prompts";

export const GEMINI_MODEL = "gemini-3.5-flash";

const TRANSCRIPTION_JSON_SCHEMA = toGeminiJsonSchema(TranscriptionSchema);
const VERIFICATION_JSON_SCHEMA = toGeminiJsonSchema(
  HandwritingVerificationSchema,
);

type InteractionResult = {
  model?: string;
  output_text?: string;
};

function isRetriable(error: unknown) {
  if (!error || typeof error !== "object") return false;
  const candidate = error as {
    status?: number;
    code?: number | string;
    name?: string;
    message?: string;
  };
  // The only abort source is the local timeout watchdog, so treat it like the
  // SDK's own timeout error.
  if (candidate.name === "AbortError") return true;
  const status = Number(candidate.status ?? candidate.code);
  return (
    status === 408 ||
    status === 429 ||
    status >= 500 ||
    /timeout|temporar|unavailable|abort/i.test(candidate.message ?? "")
  );
}

async function createInteraction(
  client: GoogleGenAI,
  parameters: Parameters<GoogleGenAI["interactions"]["create"]>[0],
): Promise<InteractionResult> {
  let lastError: unknown;
  for (let attempt = 1; attempt <= 3; attempt += 1) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 90_000);
    try {
      return (await client.interactions.create(parameters, {
        signal: controller.signal,
        timeout: 90_000,
        maxRetries: 0,
      })) as InteractionResult;
    } catch (error) {
      lastError = error;
      if (attempt === 3 || !isRetriable(error)) throw error;
      await new Promise((resolve) =>
        setTimeout(resolve, 400 * 2 ** (attempt - 1)),
      );
    } finally {
      clearTimeout(timeout);
    }
  }
  throw lastError;
}

function parseInteraction<T>(
  interaction: InteractionResult,
  model: string,
  parser: (value: unknown) => T,
  label: string,
) {
  if (interaction.model && interaction.model !== model) {
    throw new Error(
      `Model mismatch: expected ${model}, received ${interaction.model}`,
    );
  }
  if (!interaction.output_text) throw new Error(`Gemini returned no ${label}`);
  return parser(JSON.parse(interaction.output_text));
}

export function buildReview(
  transcription: Transcription,
  verification: HandwritingVerification,
  model: string,
  reviewedAt = new Date().toISOString(),
) {
  const byIndex = new Map(
    verification.problems.map((problem) => [problem.index, problem]),
  );
  if (byIndex.size !== transcription.problems.length) {
    throw new Error(
      "Gemini returned duplicate or missing handwriting verification entries",
    );
  }

  const problems = transcription.problems.map((problem, index) => {
    const check = byIndex.get(index);
    if (!check) throw new Error(`Gemini did not verify problem ${index + 1}`);
    return {
      ...problem,
      initialStudentAnswer: problem.studentAnswer,
      studentAnswer: check.studentAnswer,
      confidence: check.confidence,
      handwritingVerified: !check.needsConfirmation && check.confidence >= 0.85,
      verificationNote: check.verificationNote,
      correctAnswer: "",
      isCorrect: false,
    };
  });

  return ReviewSchema.parse(
    gradeReview({
      schemaVersion: SESSION_SCHEMA_VERSION,
      worksheetTitle: transcription.worksheetTitle,
      summary: "",
      problems,
      model,
      reviewedAt,
      ocrPasses: 2,
      verificationPasses: 2,
      gradingMethod: "local-arithmetic",
    }),
  );
}

export class WorksheetOcr {
  constructor(
    private readonly apiKey: string,
    readonly model: string,
  ) {}

  async review(imageBytes: Buffer): Promise<{
    review: Review;
    transcription: Transcription;
    verification: HandwritingVerification;
  }> {
    if (!this.apiKey) {
      throw new Error(
        "Gemini is not configured. Add GEMINI_API_KEY to the app's .env file.",
      );
    }
    const client = new GoogleGenAI({ apiKey: this.apiKey });
    const image = {
      type: "image" as const,
      data: imageBytes.toString("base64"),
      mime_type: "image/jpeg" as const,
    };
    const firstInteraction = await createInteraction(client, {
      model: this.model,
      store: false,
      input: [{ type: "text", text: TRANSCRIPTION_PROMPT }, image],
      response_format: {
        type: "text",
        mime_type: "application/json",
        schema: TRANSCRIPTION_JSON_SCHEMA,
      },
    });
    const transcription = parseInteraction(
      firstInteraction,
      this.model,
      (value) => TranscriptionSchema.parse(value),
      "transcription",
    );

    const verificationInteraction = await createInteraction(client, {
      model: this.model,
      store: false,
      input: [{ type: "text", text: verificationPrompt(transcription) }, image],
      response_format: {
        type: "text",
        mime_type: "application/json",
        schema: VERIFICATION_JSON_SCHEMA,
      },
    });
    const verification = parseInteraction(
      verificationInteraction,
      this.model,
      (value) => HandwritingVerificationSchema.parse(value),
      "handwriting verification",
    );

    const review = buildReview(transcription, verification, this.model);
    return { review, transcription, verification };
  }
}

File: ios/MathHomework/Gemini/WorksheetOCR.swift
import Foundation

/// Swift port of electron/gemini/worksheet-ocr.ts: two independent OCR-only
/// passes (structured transcription, then a handwriting check), merged and
/// graded locally.
struct WorksheetOCR {
    static let geminiModel = "gemini-3.5-flash"

    let apiKey: String
    let model: String

    struct Analysis {
        let review: Review
        let transcription: Transcription
        let verification: HandwritingVerification
    }

    func review(jpegData: Data) async throws -> Analysis {
        let client = GeminiInteractionsClient(apiKey: apiKey, model: model)

        let transcriptionText = try await client.createStructuredInteraction(
            promptText: Prompts.transcription,
            jpegData: jpegData,
            schema: GeminiSchemas.parsed(GeminiSchemas.transcription),
            label: "transcription"
        )
        let transcription = try Self.decodeValidated(
            Transcription.self, from: transcriptionText, label: "transcription"
        )
        try transcription.validate()

        let firstPassJSON = String(
            data: try JSONEncoder().encode(transcription),
            encoding: .utf8
        ) ?? "{}"
        let verificationText = try await client.createStructuredInteraction(
            promptText: Prompts.verification(firstPassJSON: firstPassJSON),
            jpegData: jpegData,
            schema: GeminiSchemas.parsed(GeminiSchemas.verification),
            label: "handwriting verification"
        )
        let verification = try Self.decodeValidated(
            HandwritingVerification.self, from: verificationText, label: "handwriting verification"
        )
        try verification.validate()

        let review = try Self.buildReview(
            transcription: transcription,
            verification: verification,
            model: model
        )
        return Analysis(review: review, transcription: transcription, verification: verification)
    }

    private static func decodeValidated<T: Decodable>(
        _ type: T.Type,
        from text: String,
        label: String
    ) throws -> T {
        guard let data = text.data(using: .utf8) else {
            throw GeminiInteractionsClient.ClientError.emptyOutput(label)
        }
        do {
            return try JSONDecoder().decode(type, from: data)
        } catch {
            throw GeminiInteractionsClient.ClientError.api(
                "Gemini returned a malformed \(label)"
            )
        }
    }

    static func buildReview(
        transcription: Transcription,
        verification: HandwritingVerification,
        model: String,
        reviewedAt: String = Timestamps.isoString()
    ) throws -> Review {
        var byIndex: [Int: HandwritingVerificationItem] = [:]
        for item in verification.problems {
            byIndex[item.index] = item
        }
        guard byIndex.count == transcription.problems.count else {
            throw ContractError.invalid(
                "Gemini returned duplicate or missing handwriting verification entries"
            )
        }

        var problems: [Problem] = []
        for (index, transcribed) in transcription.problems.enumerated() {
            guard let check = byIndex[index] else {
                throw ContractError.invalid("Gemini did not verify problem \(index + 1)")
            }
            problems.append(
                Problem(
                    number: transcribed.number,
                    expression: transcribed.expression,
                    studentAnswer: check.studentAnswer,
                    confidence: check.confidence,
                    correctAnswer: "",
                    isCorrect: false,
                    initialStudentAnswer: transcribed.studentAnswer,
                    handwritingVerified: !check.needsConfirmation && check.confidence >= 0.85,
                    verificationNote: check.verificationNote
                )
            )
        }

        let review = Review(
            schemaVersion: Contracts.sessionSchemaVersion,
            worksheetTitle: transcription.worksheetTitle,
            summary: "",
            problems: problems,
            model: model,
            reviewedAt: reviewedAt,
            ocrPasses: 2,
            verificationPasses: 2,
            gradingMethod: .localArithmetic
        )
        return Grader.gradeReview(review)
    }
}
